src/ui/camera.js

```diff
@@ -632,66 +756,102 @@ class Camera extends Evented {
-            Math.max(p0rotated.y, p1rotated.y, p2rotated.y, p3rotated.y)
-        );
-        const lowerLeft = new Point(
-            Math.min(p0rotated.x, p1rotated.x, p2rotated.x, p3rotated.x),
-            Math.min(p0rotated.y, p1rotated.y, p2rotated.y, p3rotated.y)
-        );
-
-        // Calculate zoom: consider the original bbox and padding.
-        const size = upperRight.sub(lowerLeft);
-        const scaleX = (tr.width - ((edgePadding.left || 0) + (edgePadding.right || 0) + eOptions.padding.left + eOptions.padding.right)) / size.x;
-        const scaleY = (tr.height - ((edgePadding.top || 0) + (edgePadding.bottom || 0) + eOptions.padding.top + eOptions.padding.bottom)) / size.y;
-
-        if (scaleY < 0 || scaleX < 0) {
-            warnOnce(
-                'Map cannot fit within canvas with the given bounds, padding, and/or offset.'
-            );
-            return;
-        }
-        const zoom = Math.min(tr.scaleZoom(tr.scale * Math.min(scaleX, scaleY)), eOptions.maxZoom);
 
-        // Calculate center: apply the zoom, the configured offset, as well as offset that exists as a result of padding.
-        const offset = (typeof eOptions.offset.x === 'number' && typeof eOptions.offset.y === 'number') ?
+        const worldCoords = [[p0world.x, p0world.y, 0], [p1world.x, p1world.y, 0]];
+        const aabb = Aabb.fromPoints(worldCoords);
+
+        const worldToCamera = tr.getWorldToCameraMatrix();
+        const cameraToWorld = mat4.invert(new Float64Array(16), worldToCamera);
+
+        aabb.applyTransform(worldToCamera);
+
+        const size = vec3.sub([], aabb.max, aabb.min);
+
+        const screenPadL = edgePadding.left || 0;
+        const screenPadR = edgePadding.right || 0;
+        const screenPadB = edgePadding.bottom || 0;
+        const screenPadT = edgePadding.top || 0;
+
+        const padL = eOptions.padding.left;
+        const padR = eOptions.padding.right;
+        const padT = eOptions.padding.top;
+        const padB = eOptions.padding.bottom;
+
+        const halfScreenPadX = ((edgePadding.left || 0) + (edgePadding.right || 0)) * 0.5;
+        const halfScreenPadY = ((edgePadding.top || 0) + (edgePadding.bottom || 0)) * 0.5;
+
+        const scaleX = (tr.width - (screenPadL + screenPadR + padL + padR)) / size[0];
+        const scaleY = (tr.height - (screenPadB + screenPadT + padB + padT)) / size[1];
+
+        const zoomRef = Math.min(tr.scaleZoom(tr.scale * Math.min(scaleX, scaleY)), eOptions.maxZoom);
+
+        const scaleRatio = tr.scale / tr.zoomScale(zoomRef);
+
+        aabb.setMin([
+            aabb.min[0] - (padL + halfScreenPadX) * scaleRatio,
+            aabb.min[1] - (padB + halfScreenPadY) * scaleRatio,
+            aabb.min[2]
+        ]);
+        aabb.setMax([
+            aabb.max[0] + (padR + halfScreenPadX) * scaleRatio,
+            aabb.max[1] + (padT + halfScreenPadY) * scaleRatio,
+            aabb.max[2]
+        ]);
```

Comment:
You're basically replacing the Aabb instance completely, so you could just remove `setMin`/`setMax` methods and do `new Aabb` for simplicity.